# RFM Customer Segmentation & Cohort Analysis Project

This is a RFM project which we will apply RFM Analysis and Customer Segmentation using K-Means Clustering. Naturally we will conduct Data Cleaning, Data Visualization and Exploratory Data Analysis as well as Cohort and Conduct Cohort Analysis. 

# #Determines

We are going to use the [Online Retail dataset](https://archive.ics.uci.edu/ml/datasets/Online+Retail) from the UCI Machine Learning Repository for exploratory data analysis, ***Customer Segmentation***, ***RFM Analysis***, ***K-Means Clustering*** and ***Cohort Analysis***.

This is a transnational data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail. The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers.

Feature Information:

**InvoiceNo**: Invoice number. *Nominal*, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation. 
<br>
**StockCode**: Product (item) code. *Nominal*, a 5-digit integral number uniquely assigned to each distinct product.
<br>
**Description**: Product (item) name. *Nominal*. 
<br>
**Quantity**: The quantities of each product (item) per transaction. *Numeric*.
<br>
**InvoiceDate**: Invoice Date and time. *Numeric*, the day and time when each transaction was generated.
<br>
**UnitPrice**: Unit price. *Numeric*, Product price per unit in sterling.
<br>
**CustomerID**: Customer number. *Nominal*, a 5-digit integral number uniquely assigned to each customer.
<br>
**Country**: Country name. *Nominal*, the name of the country where each customer resides.


---


First of all, to observe the structure of the data and missing values, we will use exploratory data analysis and data visualization techniques.

We are going to do descriptive analysis. Because we must understand the relationship of the features to each other and clear the noise and missing values in the data. After that, the data set will be ready for RFM analysis.

Before starting the RFM Analysis, we will do some analysis regarding the distribution of *Orders*, *Customers* and *Countries*. These analyzes will help the company develop its sales policies and contribute to the correct use of resources.

First eyecatching feature to notice is that the UK not only has the most sales revenue, but also the most customers. So we will decide whether we will continue to analyze only UK transactions or categorize countries as UK and Non-UK before we apply RFM Analysis, Customer Segmentation and K-Means Clustering.

Next, we will begin RFM Analysis, a customer segmentation technique based on customers' past purchasing behavior. 

By using RFM Analysis, one can enable companies to develop different approaches to different customer segments so that they can get to know their customers better, observe trends better, and increase customer retention and sales revenues.

We will calculate the Recency, Frequency and Monetary values of the customers in the RFM Analysis. Ultimately, we are going to create an RFM table containing these values.

In the Customer Segmentation section, we will create an RFM Segmentation Table where we segment our customers by using the RFM table. For example, we can label the best customer as "Big Spenders" and the lost customer as "Lost Customer".

We will segment the customers ourselves based on their recency, frequency, and monetary values. But can an **unsupervised learning** model do this better for us? We will use the K-Means algorithm to find the answer to this question. Then we will compare the classification made by the algorithm with the classification we have made ourselves.

Before applying K-Means Clustering, we should do data pre-processing. In this context, it will be useful to examine feature correlations and distributions. In addition, the data we apply for K-Means should be normalized.

After the above operations, we will have made cluster estimation with K-Means. We should visualize the cluster distribution by using a scatter plot. You can observe the properties of the resulting clusters with the help of the boxplot. Thus we will be able to tag clusters and interpret results.

Finally, we will do Cohort Analysis with the data you used at the beginning, regardless of the analysis we have done before. Cohort analysis is a subset of behavioral analytics that takes the user data and breaks them into related groups for analysis. This analysis can further be used to do customer segmentation and track metrics like retention, churn, and lifetime value.


# #Project Structures

- Data Cleaning & Exploratory Data Analysis
- RFM Analysis
- Customer Segmentation
- Applying K-Means Clustering
- Create Cohort and Conduct Cohort Analysis

# 1. Data Cleaning & Exploratory Data Analysis

## Importing Modules for Data Cleaning and EDA

In [2]:
# Data and Stats packages
import pandas as pd
import numpy as np

# Visualization packages
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns
plt.rcParams["axes.grid"] = False

import warnings
warnings.filterwarnings("ignore")

### i. Taking a look at the data we have

In [3]:
df_raw = pd.read_excel('Online Retail.xlsx')
df = df_raw

In [4]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [5]:
df.shape

(541909, 8)

In [6]:
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


Let's check the unique values including Nan's 

In [8]:
def first_look(i):
    val_count = df[i].value_counts(dropna=False)
    nan_i = df[i].isna().sum()
    val_count_df = pd.DataFrame(val_count)
    return print(f'FEATURE: {i}\n-------\nUNIQUE VALUES: {len(val_count)} \n\n\
    NULL VALUES: {nan_i} \n\n HEAD:\n\n{val_count_df.head(10)}\n\nTAIL:\n\n{val_count_df.tail(5)}\n')
    

In [9]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

### ii. InvoiceNo

InvoiceNo: Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation. 

In [10]:
first_look(df.columns[0])

FEATURE: InvoiceNo
-------
UNIQUE VALUES: 25900 

    NULL VALUES: 0 

 HEAD:

        InvoiceNo
573585       1114
581219        749
581492        731
580729        721
558475        705
579777        687
581217        676
537434        675
580730        662
538071        652

TAIL:

        InvoiceNo
542892          1
553849          1
553847          1
553837          1
549849          1



Now we check if we have a cancelled or somehow marked invoice number.

In [10]:
marked_invoice = df.InvoiceNo.str.extract(r'(.*\D.*)')[0].value_counts()
marked_invoice

C570867    101
C560540     57
C548460     45
C560855     41
C570828     39
          ... 
C559049      1
C544420      1
C576409      1
C547574      1
C543786      1
Name: 0, Length: 3839, dtype: int64

In [11]:
marked_invoice_df = df[df.InvoiceNo.str.extract(r'(.*\D.*)')[0].notnull()]
marked_invoice_df

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

               InvoiceDate  UnitPrice  CustomerID         Country  
141    2010-12-01 09:41:00      27.50     14527.0  United Kingdom  
154    2010-12-01 09:49:00       4.65     15311.0  United Kingdom  
235    2010-12-01 10:24:00       1.65     17548.0  United Kingdom  
236    2010-12-01 10:24:00       0.29     17548.0  United Kingdom  
237    2010-12-01 10:24:00       0.29     17548.0  United Kingdom  
...                    ...        ...         ...             ...  
540449 2011-12-09 09:57:00       0.83     14397.0  United Kingdom  
541541 2011-12-09 10:28:00     224.69     15498.0  United Kingdom  
541715 2011-12-09 11:57:00      10.95     15311.0  United Kingdom  
541716 2011-12-09 11:58:00       1.25     17315.0  United Kingdom  
541717 2011-12-09 11:58:00       1.25     17315.0  United Kingdom  

[9291 rows x 8 columns]

Since C is a marker for cancelled orders with negative values in Quentity, this sounds logical. Are there other flags than C?

In [12]:
df[df.InvoiceNo.str.extract(r'([^C^c^\d].*)')[0].notnull()]

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
299982   A563185         B  Adjust bad debt         1 2011-08-12 14:50:00   
299983   A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
299984   A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

        UnitPrice  CustomerID         Country  
299982   11062.06         NaN  United Kingdom  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

This appears to be a data entry error correction, as the unit price is assigned positive once and negative twice immediately after this entry.

We will drop all cancelled orders.

In [13]:
df[df.InvoiceNo.str.extract(r'(^\d{6}$)')[0].notnull()]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

This is unexpected. One possible reason is the data type that InvoiceNo contains. Let's check if it contains integers.

In [14]:
df[df.InvoiceNo.apply(isinstance, args = [int])]

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[532618 rows x 8 columns]

As we suspected, completed orders are stored as ordinal integer values. We will use the masking above to get completed orders and assign them to df itself.

In [15]:
df = df[df.InvoiceNo.apply(isinstance, args = [int])]

Before we proceed to the next column, let's check how many unique invoices do we have.

In [16]:
len(df.InvoiceNo.unique())

22061

### iii. StockCode

StockCode: Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product. 

In [17]:
first_look(df.columns[1])

FEATURE: StockCode
-------
UNIQUE VALUES: 4058 

    NULL VALUES: 0 

 HEAD:

        StockCode
85123A       2271
85099B       2115
22423        2022
47566        1707
20725        1595
84879        1490
22197        1426
22720        1403
21212        1370
20727        1328

TAIL:

        StockCode
79323G          1
20689           1
90025B          1
20678           1
90126C          1



In [18]:
marked_stock_df = df[df.StockCode.str.extract(r'(.*\D.*)')[0].notnull()]
marked_stock_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
45        536370      POST                              POSTAGE         3   
...          ...       ...                                  ...       ...   
541778    581578    84997C      CHILDRENS CUTLERY POLKADOT BLUE         8   
541809    581579    85099C       JUMBO  BAG BAROQUE BLACK WHITE        10   
541838    581580    84993A            75 GREEN PETIT FOUR CASES         2   
541844    581580    85049A        TRADITIONAL CHRISTMAS RIBBONS         1   
541853    581580    85049E            SCANDINAVIAN REDS RIBBONS         2   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
45     2010-12-01 08:45:00      18.00     12583.0          France  
...                    ...        ...         ...             ...  
541778 2011-12-09 12:16:00       4.15     12713.0         Germany  
541809 2011-12-09 12:19:00       1.79     17581.0  United Kingdom  
541838 2011-12-09 12:20:00       0.42     12748.0  United Kingdom  
541844 2011-12-09 12:20:00       1.25     12748.0  United Kingdom  
541853 2011-12-09 12:20:00       1.25     12748.0  United Kingdom  

[53633 rows x 8 columns]

In [19]:
marked_stock_df.StockCode.value_counts(dropna=False)

85123A    2271
85099B    2115
POST      1130
85099C     946
82494L     926
          ... 
85036a       1
90142A       1
90162A       1
35592T       1
85115B       1
Name: StockCode, Length: 1120, dtype: int64

For now we leave this marks behind and proceed to the next column.

### iv. Description

Description: Product (item) name. Nominal.

In [20]:
first_look(df.columns[2])

FEATURE: Description
-------
UNIQUE VALUES: 4207 

    NULL VALUES: 1454 

 HEAD:

                                    Description
WHITE HANGING HEART T-LIGHT HOLDER         2327
JUMBO BAG RED RETROSPOT                    2115
REGENCY CAKESTAND 3 TIER                   2019
PARTY BUNTING                              1707
LUNCH BAG RED RETROSPOT                    1594
ASSORTED COLOUR BIRD ORNAMENT              1489
NaN                                        1454
SET OF 3 CAKE TINS PANTRY DESIGN           1399
PACK OF 72 RETROSPOT CAKE CASES            1370
LUNCH BAG  BLACK SKULL.                    1328

TAIL:

                                 Description
AMBER BERTIE MOBILE PHONE CHARM            1
BIRD ON BRANCH CANVAS SCREEN               1
Amazon Adjustment                          1
????missing                                1
Dotcom sold in 6's                         1



In [21]:
description_null_df=df[df.Description.isna()]
description_null_df

InvoiceNo StockCode Description  Quantity         InvoiceDate  \
622       536414     22139         NaN        56 2010-12-01 11:52:00   
1970      536545     21134         NaN         1 2010-12-01 14:32:00   
1971      536546     22145         NaN         1 2010-12-01 14:33:00   
1972      536547     37509         NaN         1 2010-12-01 14:33:00   
1987      536549    85226A         NaN         1 2010-12-01 14:34:00   
...          ...       ...         ...       ...                 ...   
535322    581199     84581         NaN        -2 2011-12-07 18:26:00   
535326    581203     23406         NaN        15 2011-12-07 18:31:00   
535332    581209     21620         NaN         6 2011-12-07 18:35:00   
536981    581234     72817         NaN        27 2011-12-08 10:33:00   
538554    581408     85175         NaN        20 2011-12-08 14:06:00   

        UnitPrice  CustomerID         Country  
622           0.0         NaN  United Kingdom  
1970          0.0         NaN  United Kingdom  
1971          0.0         NaN  United Kingdom  
1972          0.0         NaN  United Kingdom  
1987          0.0         NaN  United Kingdom  
...           ...         ...             ...  
535322        0.0         NaN  United Kingdom  
535326        0.0         NaN  United Kingdom  
535332        0.0         NaN  United Kingdom  
536981        0.0         NaN  United Kingdom  
538554        0.0         NaN  United Kingdom  

[1454 rows x 8 columns]

UnitPrices seem to be 0.0 when Description is Null. Let's check if it is the case for all 1454 rows.

In [1]:
description_null_df[description_null_df.UnitPrice==0.0]

NameError: name 'description_null_df' is not defined

This proves our assumption and we can discard orders with 0.0 UnitPrice and NaN as Description since they are not actual orders. 

In [ ]:
df = df[df.Description.notna()]
df

### v. Quantity

Quantity: The quantities of each product (item) per transaction. Numeric. 

In [ ]:
first_look(df.columns[3])

Negative quantities are interesting, this could be an inventory correction or an order cancellation. Let's take a closer look.

In [ ]:
negative_quantity_df = df[df.Quantity<=.0]
negative_quantity_df

Not to our surprise, the unit prices are 0.0 for the orders with negative quantity. Let's check if this is the case for all 474 lines. 

In [ ]:
negative_quantity_df[negative_quantity_df.UnitPrice==.0]

We can discard these 474 rows from our dataset.

In [ ]:
df = df[df.Quantity>.0]

### vi. InvoiceDate

InvoiceDate: Invoice Date and time. Numeric, the day and time when each transaction was generated. 

In [ ]:
first_look(df.columns[4])

In [ ]:
df.InvoiceDate.max()

In [ ]:
df.InvoiceDate.min()

In [ ]:
order_by_year = pd.DatetimeIndex(df['InvoiceDate']).year.value_counts()
order_by_year

In [ ]:
order_by_month = pd.DatetimeIndex(df['InvoiceDate']).month.value_counts()
order_by_month

In [ ]:
order_by_hour = pd.DatetimeIndex(df['InvoiceDate']).hour.value_counts()
order_by_hour

These temporal trends can have a significant impact on customer behavior. For now, we will not add new features such as month and hour of purchase, but before we create clusters, we will look at this issue again.

### vii. UnitPrice

UnitPrice: Unit price. Numeric, Product price per unit in sterling. 

In [ ]:
first_look(df.columns[5])

In [ ]:
df.UnitPrice.max()

In [ ]:
df.UnitPrice.min()

In [ ]:
df.UnitPrice.describe()

In [ ]:
df[df.UnitPrice==.0]

We are going to drop rows that have 0.0 UnitPrice.

In [ ]:
df = df[df.UnitPrice>0.0]

### viii. CustomerID

CustomerID: Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer. 

In [ ]:
first_look(df.columns[6])

We have a large amount of null values in the "CustomerID" column. Since this is a unique number, we have limited options to populate these values and we will most likely discard them. 
Let's take a closer look.

In [ ]:
df[df.CustomerID.isna()]

All characteristics except the customer ID seem to be concrete. On the other hand, the customer ID is a key feature from the perspective of RMF analysis, and orders without it could lead to biases in the analysis. Hence we are going to discard these rows in our dataset. For future analyses, we will save a version of the dataset that also includes NaN customer ID values, but our main dataset will be cleaned of these NaN values.

In [ ]:
df_with_na_customers = df
df = df[df.CustomerID.notna()]

### iX. Country

Country: Country name. Nominal, the name of the country where each customer resides.

In [ ]:
first_look(df.columns[7])

In [ ]:
country_dist = pd.DataFrame(df.Country.value_counts())
country_dist

In [ ]:
df[df.Country=='Unspecified']

In 241 orders Countries is not specified. Yet, they seem valid and we will keep them.  

In [ ]:
plt.figure(figsize=(18,10))
sns.barplot(y=country_dist.index,x='Country',data=country_dist)

As we can see, the UK has the largest share of our dataset. We have two options here: The first is to consider only orders from the UK, or we can consider two types of countries, domestic and international. We choose the second option in order not to lose valuable features of customer behavior. Yet, we leave this transformation to the next step.

### X. Handling Duplications

In [ ]:
df[df.duplicated()]

We have 5192 rows that are duplicated, and since our orders are unique, we should discard these duplicates.

In [ ]:
df = df.drop_duplicates(keep='first')
df

### xi. Exploring the Orders


    (1) Unique number of InvoiceNo per customer

In [ ]:
custmr_invoice_unique = pd.DataFrame(df.groupby('CustomerID').InvoiceNo.unique())
custmr_invoice_unique['Total_Invoices']=custmr_invoice_unique.InvoiceNo.apply(lambda x:len(x))

In [ ]:
custmr_invoice_unique.sort_values(['Total_Invoices'], ascending=False, inplace=True )
custmr_invoice_unique

In [ ]:
avg_custmr_invoice = custmr_invoice_unique.Total_Invoices.sum()/custmr_invoice_unique.shape[0]
print(f'Our customers issued {avg_custmr_invoice:.2f} unique orders on an average.')

    (2) Average number of unqiue items per order

In [ ]:
invoice_item_unique = pd.DataFrame(df.groupby('InvoiceNo').StockCode.unique())
invoice_item_unique['Total_Items'] = invoice_item_unique.StockCode.apply(lambda x:len(x))
invoice_item_unique.sort_values(['Total_Items'], ascending=False, inplace=True)
invoice_item_unique

In [ ]:
avg_invoice_item = invoice_item_unique.Total_Items.sum()/invoice_item_unique.shape[0]
print(f'We\'ve had {avg_invoice_item:.1f} unique items of order on an average.')

    (3) Average number of unqiue items per customer

In [ ]:
custmr_item_unique = pd.DataFrame(df.groupby('CustomerID').StockCode.unique())
custmr_item_unique['Total_Items'] = custmr_item_unique.StockCode.apply(lambda x:len(x))
custmr_item_unique.sort_values(['Total_Items'], ascending=False, inplace=True)
custmr_item_unique

In [ ]:
avg_custmr_item = custmr_item_unique.Total_Items.sum()/custmr_item_unique.shape[0]
print(f'Our customers ordered {avg_custmr_item:.1f} unique items on an average.')

### xii. Exploring Customers by Country

    (1) Total revenue per country

In [ ]:
df['Revenue'] = df.Quantity * df.UnitPrice
df.head()

In [ ]:
country_revenue = pd.DataFrame(df.groupby('Country').Revenue.sum())
country_revenue.sort_values('Revenue', ascending=False, inplace=True)

In [ ]:
plt.figure(figsize=(18,10))
sns.barplot(x='Revenue', y=country_revenue.index, data=country_revenue)

In [ ]:
print(f'United Kingdom posesses % {100*country_revenue.Revenue[0]/sum(country_revenue.Revenue[:]):.2f} of all revenues we have.')

    (2) Visualizing number of customer per country

In [ ]:
country_customer = pd.DataFrame(df.groupby('Country').CustomerID.sum())
country_customer.sort_values('CustomerID', ascending=False, inplace=True)

In [ ]:
plt.figure(figsize=(18,10))
sns.barplot(x='CustomerID', y=country_customer.index, data=country_customer)
plt.xlabel('Number of Customers')

In [ ]:
print(f'% {100*country_customer.CustomerID[0]/sum(country_customer.CustomerID[:]):.2f} of all our customers resides in United Kingdom.')

    (3) The most popular products by country

In [ ]:
county_products_quantity = pd.DataFrame(df.groupby(['Country','StockCode']).Quantity.sum())
country_products_revenue = pd.DataFrame(df.groupby(['Country','StockCode']).Revenue.sum())
country_product = pd.merge(county_products_quantity, country_products_revenue, on=['Country','StockCode'])
country_product

In [ ]:
country_product.groupby('Country').Revenue.sum().sort_values(ascending=False)

In [ ]:
def country_trend(country_x):
    country_df = country_product[country_product.index.isin([country_x], level=0)]
    country_df = country_df.sort_values('Quantity', ascending=False)
    return country_df

In [ ]:
country_trend('United Kingdom').head(10)

In [ ]:
country_trend('Netherlands').head(10)

    (4) Transforming Country as UK and Non-UK

In [ ]:
df['Country'] = df.Country.apply(lambda x:1 if x=='United Kingdom' else 0)

In [ ]:
df.sample(10)

### We will continue analyzing transactions with customer segmentation.

# 2. RFM Analysis

In the age of the internet and e-commerce, companies that do not expand their businesses online or utilize digital tools to reach their customers will run into issues like scalability and a lack of digital precsence. An important marketing strategy e-commerce businesses use for analyzing and predicting customer value is customer segmentation. Customer data is used to sort customers into group based on their behaviors and preferences.

**[RFM](https://www.putler.com/rfm-analysis/) (Recency, Frequency, Monetary) Analysis** is a customer segmentation technique for analyzing customer value based on past buying behavior. RFM analysis was first used by the direct mail industry more than four decades ago, yet it is still an effective way to optimize your marketing.
<br>
<br>
Our goal in this Notebook is to cluster the customers in our data set to:
 - Recognize who are our most valuable customers
 - Increase revenue
 - Increase customer retention
 - Learn more about the trends and behaviors of our customers
 - Define customers that are at risk

We will start with **RFM Analysis** and then compliment our findings with predictive analysis using **K-Means Clustering Algorithms.**

- RECENCY (R): Time since last purchase
- FREQUENCY (F): Total number of purchases
- MONETARY VALUE (M): Total monetary value


Benefits of RFM Analysis

- Increased customer retention
- Increased response rate
- Increased conversion rate
- Increased revenue

RFM Analysis answers the following questions:
 - Who are our best customers?
 - Who has the potential to be converted into more profitable customers?
 - Which customers do we need to retain?
 - Which group of customers is most likely to respond to our marketing campaign? 

### i. Import Libraries

In [ ]:
import datetime as dt
from datetime import timedelta

### ii. Review DataFrame

In [ ]:
df.head()

In [ ]:
rfm_df = df

### iii. Recency: Days since last purchase

    (1) Choosing a date as a point of reference to evaluate how many days ago was the customer's last purchase.

In [ ]:
rfm_df['Date']=rfm_df.InvoiceDate.dt.date

In [ ]:
rfm_df.head()

In [ ]:
ref_date = rfm_df.Date.max() + timedelta(days=2)
ref_date

     (2) Grouping by CustomerID and checking the last date of purchase

In [ ]:
customer_recency = pd.DataFrame(rfm_df.groupby('CustomerID', as_index=False).Date.max())
customer_recency.head()

    (3) Calculating Recency

In [ ]:
customer_recency['Recency'] = ref_date - customer_recency.Date
customer_recency

In [ ]:
customer_recency['Recency'] = pd.to_numeric(customer_recency['Recency'].dt.days.astype('int64'))

In [ ]:
customer_recency.drop(columns=['Date'], inplace=True, axis=1)

In [ ]:
customer_recency

    (4) Plot RFM distributions

In [ ]:
plt.figure(figsize=(10,8))
sns.scatterplot(data=customer_recency.Recency)

In [ ]:
plt.figure(figsize=(10,8))
sns.distplot(customer_recency.Recency, kde=False, rug=True)

### iv. Frequency: Number of purchases

To calculate how many times a customer purchased something, we need to count how many invoices each customer has.

    (1) Calculating the frequency of purchases

In [ ]:
customer_frequency = rfm_df.groupby('CustomerID',as_index=False)['InvoiceNo'].nunique()

In [ ]:
customer_frequency.rename(columns={'InvoiceNo':'Frequency'}, inplace=True)
customer_frequency.head()

    (2) Plot RFM distributions

In [ ]:
plt.figure(figsize=(10,8))
sns.scatterplot(data=customer_frequency.Frequency)

In [ ]:
plt.figure(figsize=(10,8))
sns.distplot(customer_frequency.Frequency, kde=False)

### v. Monetary: Total amount of money spent

    (1) Calculating the "Monetary" by customers

In [ ]:
customer_monetary = rfm_df.groupby('CustomerID',as_index=False)['Revenue'].sum()
customer_monetary.rename(columns={'Revenue':'Monetary'}, inplace=True)
customer_monetary.head()

    (2) Plot Monetary distributions

In [ ]:
plt.figure(figsize=(10,8))
sns.scatterplot(data=customer_monetary.Monetary)

In [ ]:
plt.figure(figsize=(10,8))
sns.distplot(customer_monetary.Monetary, kde=False)

### vi. Merging the recency, frequency and motetary dataframes

In [ ]:
customer_rfm = pd.merge(pd.merge(customer_recency, customer_frequency, on='CustomerID'), customer_monetary, on='CustomerID')
customer_rfm.head()

# 3. Customer Segmentation with RFM Scores

## Calculating RFM Scoring

The simplest way to create customer segments from an RFM model is by using **Quartiles**. We will assign a score from 1 to 4 to each category (Recency, Frequency, and Monetary) with 4 being the highest/best value. The final RFM score is calculated by combining all RFM values. For Customer Segmentation, we will use the customer_rfm data set resulting from the RFM analysis.


### i. Creating the RFM Segmentation Table


    (1) Dividing the customer_rfm into quarters

In [ ]:
customer_rfm.describe()

    (2) RFM scoring

In [ ]:
def recency_scoring(rfm):
    if rfm.Recency <= 19.0:
        recency_score = 4
    elif rfm.Recency <= 52.0:
        recency_score = 3
    elif rfm.Recency <= 143.0:
        recency_score = 2
    else:
        recency_score = 1
    return recency_score
customer_rfm['R'] = customer_rfm.apply(recency_scoring, axis=1)
customer_rfm.head()

In [ ]:
def frequency_scoring(rfm):
    if rfm.Frequency >= 10.0:
        frequency_score = 4
    elif rfm.Frequency >= 6.0:
        frequency_score = 3
    elif rfm.Frequency <= 3.0:
        frequency_score = 2
    else:
        frequency_score = 1
    return frequency_score
customer_rfm['F'] = customer_rfm.apply(frequency_scoring, axis=1)
customer_rfm.head()

In [ ]:
def monetary_scoring(rfm):
    if rfm.Monetary >= 1660.0:
        monetary_score = 4
    elif rfm.Monetary >= 668.0:
        monetary_score = 3
    elif rfm.Monetary <= 306.0:
        monetary_score = 2
    else:
        monetary_score = 1
    return monetary_score
customer_rfm['M'] = customer_rfm.apply(monetary_scoring, axis=1)
customer_rfm.head()

In [ ]:
def rfm_scoring(customer):
    return str(int(customer['R'])) + str(int(customer['F'])) + str(int(customer['M']))


customer_rfm['RFM_score'] = customer_rfm.apply(rfm_scoring, axis=1)
customer_rfm.head()

In [ ]:
segments = {'Customer Segment':['Champions','Loyal Customers','Potential Loyalist', 'Recent Customers', 'Customers Needing Attention', 'Still Got Hope', 'Need to Get Them Back','Lost', 'Give it a Try'],\
            'RFM':['(3|4)-(3|4)-(3|4)', '(2|3|4)-(3|4)-(1|2|3|4)', '(3|4)-(2|3)-(1|2|3|4)', '(4)-(1)-(1|2|3|4)', '(2|3)-(2|3)-(2|3)', '(2|3)-(1|2)-(1|2|3|4)', '(1|2)-(3|4)-(3|4)', '(1|2)-(1|2)-(1|2)','(1|2)-(1|2|3)-(1|2|3|4)']}
pd.DataFrame(segments)

In [ ]:
def cathegorizer(rfm):
    if (rfm[0] in ['3', '4']) & (rfm[1] in ['3', '4']) & (rfm[2] in ['3', '4']):
        rfm = 'Champions'
        
    elif (rfm[0] in ['2', '3', '4']) & (rfm[1] in ['3', '4']) & (rfm[2] in ['1', '2', '3', '4']):
        rfm = 'Loyal Customers'
        
    elif (rfm[0] in ['3', '4']) & (rfm[1] in ['2', '3']) & (rfm[2] in ['1', '2', '3', '4']):
        rfm = 'Potential Loyalist'
    
    elif (rfm[0] in ['4']) & (rfm[1] in ['1']) & (rfm[2] in ['1', '2', '3', '4']):
        rfm = 'Recent Customers'
    
    elif (rfm[0] in ['2', '3']) & (rfm[1] in ['2', '3']) & (rfm[2] in ['2', '3']):
        rfm = 'Customers Needing Attention'
    
    elif (rfm[0] in ['2', '3']) & (rfm[1] in ['1', '2']) & (rfm[2] in ['1', '2', '3', '4']):
        rfm = 'Still Got Hope'
    
    elif (rfm[0] in ['1', '2']) & (rfm[1] in ['3', '4']) & (rfm[2] in ['3', '4']):
        rfm = 'Need to Get Them Back'
                
    elif (rfm[0] in ['1', '2']) & (rfm[1] in ['1', '2']) & (rfm[2] in ['1', '2']):
        rfm = 'Lost'
    
    elif (rfm[0] in ['1', '2']) & (rfm[1] in ['1', '2', '3']) & (rfm[2] in ['1', '2', '3', '4']):
        rfm = 'Give it a Try'
    
    return rfm 

In [ ]:
customer_rfm['Cathegory'] = customer_rfm.RFM_score.apply(cathegorizer)

In [ ]:
customer_rfm.Cathegory.value_counts(dropna=False, normalize=True)

In [ ]:
customer_rfm.head()

## Plot RFM Segments

1. Create your plot and resize it.

In [ ]:
segmented = pd.DataFrame(customer_rfm.Cathegory.value_counts(dropna=False).sort_values(ascending=False))
segmented.reset_index(inplace=True)
segmented.rename(columns={'index':'Segments', 'Cathegory':'Customers'}, inplace=True)

In [ ]:
segmented

In [ ]:
plt.figure(figsize=(19, 8))
sns.barplot(data=segmented, x='Segments', y='Customers', palette='Oranges_r')

An example for using customer segmentation categories found [here](http://www.blastam.com/blog/rfm-analysis-boosts-sales). Since we applied cathegorisation that suits our needs we can formulate different marketing strategies and approaches for customer engagement for each type of customer.

# 3. Applying K-Means Clustering

Now that we have our customers segmented into 9 different categories, we can gain further insight into customer behavior by using predictive models in conjuction with out RFM model.
Possible algorithms include **Logistic Regression**, **K-means Clustering**, and **K-nearest Neighbor**. We will go with [K-Means](https://towardsdatascience.com/understanding-k-means-clustering-in-machine-learning-6a6e67336aa1) since we already have our distinct groups determined. K-means has also been widely used for market segmentation and has the advantage of being simple to implement.

## Data Pre-Processing and Exploring

In [ ]:
country_df = pd.DataFrame(rfm_df.groupby('CustomerID').Country.unique())
country_df['Country'] = country_df.Country.apply(lambda x:x[0])

In [ ]:
kmeans_df = customer_rfm.set_index('CustomerID')[['Recency','Frequency','Monetary']]
kmeans_df = kmeans_df.merge(country_df, on='CustomerID')
kmeans_df

### i. Defining and Plotting Feature Correlations

Creating Heatmap and evaluating the the correlation between the features

In [ ]:
kmeans_df.corr()

In [ ]:
plt.figure(figsize=(18,10))
sns.heatmap(data = kmeans_df.corr(), annot=True)

### ii. Visualizing Feature Distributions

To get a better understanding of the dataset, we can costruct a scatter matrix of each of the three features.

In [ ]:
plt.figure(figsize=(18,10))
sns.pairplot(data = kmeans_df)

### iii. Data Normalization

1. We will use the logarithm method to normalize the values in a column.

In [ ]:
kmeans_df['Recency'] = (kmeans_df.Recency + 0.1).apply(np.log)
kmeans_df['Frequency'] = (kmeans_df.Frequency + 0.1).apply(np.log)
kmeans_df['Monetary'] = kmeans_df.Monetary.apply(np.log)
kmeans_df.sample(10)

2. Ploting normalized data with scatter matrix or pairplot.

In [ ]:
plt.figure(figsize=(18,10))
sns.pairplot(data = kmeans_df, hue='Country')

In [ ]:
kmeans_df.corr()

In [ ]:
sns.heatmap(kmeans_df.corr(), annot=True)

## K-Means Implementation

For k-means, we have to set k to the number of clusters we want, but figuring out how many clusters is not obvious from the beginning. We will try different cluster numbers and check their [silhouette coefficient](http://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_silhouette_analysis.html). The silhouette coefficient for a data point measures how similar it is to its assigned cluster from -1 (dissimilar) to 1 (similar). 
<br>
<br>
**Note**: K-means is sensitive to initializations because they are critical to qualifty of optima found. Thus, we will use smart initialization called "Elbow Method".

In [ ]:
from sklearn.cluster import KMeans
from sklearn.metrics import adjusted_rand_score
from yellowbrick.cluster import KElbowVisualizer
from sklearn.metrics import silhouette_samples, silhouette_score

### i. Define the Optimal Number of Clusters

In [ ]:
from pyclustertend import hopkins
hopkins(kmeans_df, kmeans_df.shape[0])

Our Hopkins value is too close to 0 which means we have **a dataset which has a quite small tendency to clustering**. We need to further analyse with silouhette scores whether our data has a tendency to clustering or not.

[Yellow Brick Elbow](https://www.scikit-yb.org/en/latest/api/cluster/elbow.html)

In [ ]:
plt.rcParams['figure.facecolor'] = 'white'
model = KMeans()
visualizer = KElbowVisualizer(model, k=(4,12))

visualizer.fit(kmeans_df)        # Fit the data to the visualizer
visualizer.show() 

According to YellowBrick Elbow method we should apply 7 n_clusters. We will also check Classic Elbow Methode.

[Classic Elbow Method](https://en.wikipedia.org/wiki/Elbow_method_(clustering) 

In [ ]:
import pickle

ssd = []

K = range(1,10)

matrix = kmeans_df.values

for k in K:
    model = KMeans(n_clusters=k)
    model.fit(matrix)
    ssd.append(model.inertia_)

In [ ]:
plt.plot(K, ssd, "bx-")
plt.xlabel("Different k values")
plt.ylabel("inertia-error")
plt.title("Elbow Method")

[Silhouette Coefficient](http://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_silhouette_analysis.html)

In [ ]:
import matplotlib.cm as cm
X = kmeans_df.values

for n_clusters in range(2,11):

    clusterer = KMeans(n_clusters=n_clusters, random_state=10)
    cluster_labels = clusterer.fit_predict(X)

    silhouette_avg = silhouette_score(X, cluster_labels)
    print("For n_clusters =", n_clusters,
          "The average silhouette_score is :", silhouette_avg)

    sample_silhouette_values = silhouette_samples(X, cluster_labels)

    y_lower = 10

The silhouette_score is descending as the granulirity arises. This supports our first finding regarding tendency to clustering which we found too small by Hopkins methode.

For the sake of tendency to clustering we will pick ***4 n_clusters*** which also has an acceptable elbow score. (Yellowbrick Elbow method reccomends 7 clusters however silhoutte score is too low for 7 n_clousters.) 

### ii. Model Fitting

Fitting the K-Means Algorithm with the optimal number of clusters we decided and saving the model to disk.

In [ ]:
kmeans = KMeans(n_clusters = 4).fit(kmeans_df)
labels = kmeans.labels_
kmeans_df['Labels'] = kmeans.labels_
kmeans_df.head()

### iii. Visualizing the Clusters

    (1) Creating a scatter plot and select cluster centers

In [ ]:
plt.scatter(kmeans_df.iloc[:,0], kmeans_df.iloc[:,1], c = labels, s = 50, cmap = "viridis")
plt.scatter(kmeans.cluster_centers_[:, 0], kmeans.cluster_centers_[:, 1], s=200, c='red',alpha=0.5, label = 'Centroids')

    (2) Visualizing Cluster Id vs Recency, Cluster Id vs Frequency and Cluster Id vs Monetary using Box plot. Also evaluate the results. 

In [ ]:
sns.boxplot(x='Labels', y='Recency', data=kmeans_df)

In [ ]:
sns.boxplot(x='Labels', y='Frequency', data=kmeans_df)

In [ ]:
sns.boxplot(x='Labels', y='Monetary', data=kmeans_df)

### iv. Concluding remarks for K-Means Clustering

In [ ]:
kmeans_df.head(10)

In [ ]:
kmeans_df.Labels.value_counts(normalize=True)

**Conclusion**

As Silouhette Analysis and Hopkins method claimed our clusters out of K-Means Clustering are not clearly seperated. On the other hand we obtained 4 groups of customers regarding their RFM features and also location. 

- Cluster 0 : The first cluster could be called "Best Customers" (39.6% of all customers) as they purchase recently, frequent buyers, and spent the most. 

- Cluster 1 : Second cluster is an obvious 'New Commers' (35.3% of all customers) as they purchase recently with lowest frequency and low spending.

- Cluster 2 : Third cluster is highly critical and we may call them  'Alerting Precious' (5.3% of all customers) as they used to purchase with very high spending at a high frequency however their recency is too low. 

- Cluster 3 : Fourth class could be called "Promising Loyals" (19.5% of all customers) as they purchase very recently, and spent high however with a low frequency.

- What is missing : A group of customers who bought a very long time ago and have not shown up since, with low frequency and too little spending. 

**Annotation:**

Limitations of K-means clustering:

1. There is no assurance that it will lead to the global best solution.
2. Can't deal with different shapes(not circular) and consider one point's probability of belonging to more than one cluster.

These disadvantages of K-means show that for many datasets (especially low-dimensional datasets), it may not perform as well as we might hope.

# 5. Creating Cohort & Conducting Cohort Analysis
[Cohort Analysis](https://medium.com/swlh/cohort-analysis-using-python-and-pandas-d2a60f4d0a4d) is specifically useful in analyzing user growth patterns for products. In terms of a product, a cohort can be a group of people with the same sign-up date, the same usage starts month/date, or the same traffic source.
Cohort analysis is an analytics method by which these groups can be tracked over time for finding key insights. This analysis can further be used to do customer segmentation and track metrics like retention, churn, and lifetime value.

Since we will be performing Cohort Analysis based on transaction records of customers, the columns we will be dealing with mainly:
- Invoice Data
- CustomerID
- Price
- Quantity

## Feature Engineering

### i. Extracting the Month of the Purchase

In [ ]:
cohort_df = rfm_df
cohort_df.head()

In [ ]:
cohort_df.info()

Now we will use the function created above to convert all the invoice dates into respective month date format.

In [ ]:
def first_of_month(date_x):    
    dt_formatted = dt.datetime.strptime(str(date_x), '%Y-%M-%D %H:%M:%S')
    return dt.datetime(dt_formatted.year, dt_formatted.month, 1)

In [ ]:
cohort_df['InvoiceMonth'] = cohort_df['InvoiceDate'].apply(get_month)
cohort_df['CohortMonth'] = cohort_df.groupby('CustomerID')['InvoiceMonth'].transform('min')

In [ ]:
cohort_df[['InvoiceDate','InvoiceMonth', 'CohortMonth']].head()

### ii. Calculating time offset in Months i.e. Cohort Index:
Calculating time offset for each transaction will allows us to report the metrics for each cohort in a comparable fashion.
First, we will create 4 variables that capture the integer value of years, months for Invoice and Cohort Date. 

In [ ]:
def date_granule(dfx, colx):
    year = dfx[colx].dt.year
    month = df[colx].dt.month
    day = df[colx].dt.day
    return year, month, day

In [ ]:
invoice_year, invoice_month, invoice_day = date_granule(cohort_df, 'InvoiceMonth')
cohort_year, cohort_month, cohort_day = date_granule(cohort_df, 'CohortMonth')

In [ ]:
diff_year = invoice_year - cohort_year 
diff_month = invoice_month - cohort_month

cohort_df['CohortIndex'] = diff_year * 12 + diff_month + 1

In [ ]:
cohort_df[['CustomerID', 'InvoiceDate', 'InvoiceMonth', 'CohortMonth', 'CohortIndex']].sample(10)

## Creating 1st Cohort: User number & Retention Rate

### i. Pivot Cohort and Cohort Retention

In [ ]:
cohort_first = cohort_df.groupby(['CohortMonth', 'CohortIndex'])['CustomerID'].nunique().reset_index()
pivot_first = cohort_first.pivot(index='CohortMonth', columns='CohortIndex', values='CustomerID').round(1)
pivot_first

In [ ]:
sizes = pivot_first.iloc[:,0]
retention = pivot_first.divide(sizes, axis=0).round(3)*100
retention

### ii. Visualizing analysis of cohort 1

In [ ]:
plt.figure(figsize=(12,8))
plt.title('Retention Rates')
sns.heatmap(data=retention, annot=True, cmap='Greens', fmt='g')

## Create the 2nd Cohort: Average Quantity Sold

### i. Pivot Cohort

In [ ]:
cohort_second = cohort_df.groupby(['CohortMonth', 'CohortIndex'])['Quantity'].mean().reset_index()
pivot_second = cohort_second.pivot(index='CohortMonth', columns='CohortIndex', values='Quantity').round(1)
pivot_second

### ii. Visualizing Cohort-2

In [ ]:
plt.figure(figsize=(12,8))
plt.title('Average Quantity')
sns.heatmap(data=pivot_second, annot=True, cmap='Greens', fmt='g')

## Create the 3rd Cohort: Average Sales


### i. Pivot Cohort

In [ ]:
cohort_third = cohort_df.groupby(['CohortMonth', 'CohortIndex'])['Revenue'].mean().reset_index()
pivot_third = cohort_third.pivot(index='CohortMonth', columns='CohortIndex', values='Revenue').round(1)
pivot_third

### ii. Visualizing Cohort-3

In [ ]:
plt.figure(figsize=(12,8))
plt.title('Retention Rates')
sns.heatmap(data=pivot_third, annot=True, cmap='Greens')

For e-commerce organisations, cohort analysis is a unique opportunity to find out which clients are the most valuable to their business. By performing Cohort analysis we can get answers to following questions:

- How much effective was a marketing campaign held in a particular time period?
- Did the strategy employed to improve the conversion rates of Customers worked?
- Should I focus more on retention rather than acquiring new customers?
- Are my customer nurturing strategies effective?
- Which marketing channels bring me the best results?
- Is there a seasoanlity pattern in Customer behahiour?

## 6. COCLUSION

During the project, we performed three different types of analyses independently. 
All three analysis methods offer some specific advantages and also have limitations. 
In this project, we were able to see in which area and to what extent we should use these methods.

- First, we used RFM segmentation to divide our customers into insightful clusters. 
What makes RFM analysis attractive is the flexibility it offers. We can specify the segmentation in terms of our business needs.
On the other hand, it is susceptible to user-induced biases.

- Then we applied K-Means clustering to see the unsupervised machine learning results as clusters.
Since our dataset does not have a significant tendency to cluster, K-Means clustering did not produce a satisfactory clustering.

- Finally, we used cohort analysis to gain new insights from the data related to time-based cohorts. 
The strength of cohort analysis is its elasticity, as it provides specific analysis based on business needs.
